doc/requirements/Requirements.md

```diff
@@ -0,0 +1,704 @@
+This also mean that no function can be defined between criteria and states.
+I.e.: parameter values MAY depend on previous selection criterion states.</nb>
+<why>This is what the reference implementation does.</why>
+
+TODO sequence aware domains
+
+# (de)serialization
+
+## Philosophy
+Serialization and deserialization are meant to support destruction recovery and
+configuration deployment.
+
+These are the same requirements than for a database, it needs to be able to save
+its state and restore for backup, deployment, reboot...
+
+## Definition
+PF data includes:
+
+- parameters tree
+- configurations:
+    - selection rule
+    - parameter/value couples
+- domain:
+    - list of associated configurations
+    - order of priority
+
+## Requirement
+The PF data **MUST** be deserializable.
+<why>Otherwise a PF instance could only be created empty and then be filled by
+the tuning interface. The reference implementation supports it.</why>
+
+The PF data **SHOULD** be deserializable from a config file.
+<why>This is usually how program configuration are stored. The reference
+implementation supports it.</why>
+
+The PF data **SHOULD** be serializable.
+<why>In order to save a PF instance state and restore it later. This achieve
+destruction recovery. The reference implementation supports it.</why>
+
+The PF data **SHOULD** be serializable/deserializable by parts. <why>For easier
+configuration management: for versioning; for selecting only wanted parts of a
+complete configuration.</why>
+
+**TODO**: XML ?
+
+**TODO**: get the binary content of an element.
+
+# Introspection
+## Philosophy
+In order to debug
+
+## Requirements
+User **SHOULD** be able to inspect PF data.
+<why>To offer run time debugging.</why>
+This includes:
+
+- listing
+    - domains
+    - configurations of a domains
+    - parameters
+    - a domain's associated parameters
+- getting their properties. Including:
+    - parameters values, min, max, size...
+
+PF **MAY** offer pretty print of data. Including:
+
+- printing parameter value in decimal
+    <why>For human readability</why>
+- pretty print parameter tree (such as the Unix tree command for files)
+    <why>In order to ease runtime debug.</why>
+
+Users **SHOULD** be able to modify rogue parameters through the native API at
+all time.
+<why>Otherwise, a rogue parameter is of no use.</why>
+<ko>In the reference implementation, under certain conditions, this is not
+possible (tuning mode)</ko>
+
+# Tuning
+## Philosophy
+Tuning is the ability to modify the PF data structure at runtime.
+
+As the PF might model a complex system with its dynamic parameter value engine
+(rule based in the default implementation), its behaviour might be hard to
+understand and should be easily modified not correct.
+
+To address this need, a fast modify-update-test cycle should be possible.
+
+## Requirements
+Users **SHOULD** be able to modify the PF inference engine behaviour (rules,
+configuration...) with minimal effort.
```

Comment:
Of course not. It means that the engine modify-install-test cycle should be as minimal as possible. Can you think of a better way to formulate it ?